ipam/allocator.go

```diff
@@ -397,20 +402,34 @@ func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error)
 		return nil, err
 	}
 
-	for _, nw := range a.getPredefineds(as) {
+	pdl := a.getPredefineds(as)
+
+	aSpace.Lock()
+	defer aSpace.Unlock()
+
+	c := a.asIndices[as]
+
+	// Defensive and redundant unless there is a bug
+	if c < 0 {
+		c = 0
+	}
+
+	for i := 0; i < len(pdl); i, c = i+1, c+1 {
```

Comment:
actually I would use the `%`.

if you remove line 430 and at line 413 you put: `c = (c+1)% len(pdl)` you can remove also the if 412-415 (not sure if you have to handle the case of len(pdl) == 0 or is already handled before so cannot happen.

at the end of the loop or on the top then you can still use the same modulo format for the increment

Reply:
Go doesn't have a comma operator, so you can't do `i++,c++`. 

Re: removing line 430 -- that's the line that saves the cursor so that it's state is kept around for the next call.  Can't remove it. 

Yes, could use modulo, but I was coding defensively against the possibility that the list could change size.  As such there had to be *something* that kept `c <= len(pdf)` in the *first* iteration.  Yes, it could have been written as:
```
for i, c := 0, c%len(pdl); i < len(pdl); i, c = i+1, (c+1)%len(pdl) {
...
```
But that looks even messier.  I was striving for clarity.  (and if performance were a concern: these days a small branch usually costs much less than a modulo operation)

Reply:
sorry I meant to save the value of c and not the next value, so that when you start the for loop you can init it with (c+1)% new length.
Also I was thinking that because there is a lock the length should remain fixed right?

Reply:
Ah, ok.  I understand now.

Re: the lock ... I was really referring to a change in the length of the pools between calls to getPredefinedPool() ...  In practice that currently never happens because the list of pools gets set once at initialization time and then left alone.  So, again, much more defensive than it probably needs to be.

Looking at the code, I realize that I'd assumed that getPredefineds() returned a copy of the list for some good reason like being called in multiple places or being called in the face of changes.  But it only ever gets invoked in getPredefinedPool.  So maybe it is worth getting rid of entirely.

Actually, above there was a specific reason that I *didn't* do `c := (a.asIndices[as]+1) % len(pdl)` or `for i, c := 0, (a.asIndices[as]+1)%len(pdl) ; ... {`   It was even more defensiveness.   If the length of the predefined pools were 0 for any reason (i.e. the list was empty) then this would through an integer exception.

Reply:
Given the above, it's apparent that the code could be clearer.   Would the following be ok?  (comments not to go in real code)

```
// REMOVE getPredefineds() entirely
...
aSpace.Lock()
defer aSpace.Unlock()

predefined := a.predefined[as]   // avoid map lookup per reference
for i, pi := 0, a.predefinedStartIdx[as]; i < len(predefined); i, pi := i+1,pi+1 {
    if pi >= len(predefined) {    // no test for pi < 0, but can only happen with corruption and we'll get a noisy panic if it occurs
        pi = 0
    } 
    nw := predefined[pi]
    ...
        a.predefinedStartIdx[as] = pi + 1
        return nw, nil
}
```